### PR TITLE
Refactor CorDapp loading into a separate module

### PR DESCRIPTION
**Pull request by Clintonio**
Why: Mostly because this is groundwork for loading contract attachments into the attachment store and necessary for V1. But it also cleans up the AbstractNode slightly, and this file gets changed so often that I have to keep rebasing every day, putting this ground work in first will save a lot of time.

What: Mostly splitting out the cordapp loading code into a separate class and module.  

Note: I may revert the flow classpath tests and changes to the app context since they're not entirely relevant.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package net.corda.node.internal.classloading
+
+inline fun <reified A : Annotation> Class<*>.requireAnnotation(): A {
```
Comment (rnicoll) [inline node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt]: Could you add a `@file:JvmName` annotation so it doesn't create a UtilsKt class please.

File: node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import java.net.JarURLConnection
+import java.net.URI
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
+import java.util.stream.Collectors
+import kotlin.reflect.KClass
+
+/**
+ * Handles Cordapp loading and classpath scanning
```
Comment (rnicoll) [inline node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt]: "CorDapp" please

File: node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+        // Rather than looking in the plugins directory, figure out the classpath for the given package and scan that
+        // instead. This is used in tests where we avoid having to package stuff up in jars and then having to move
+        // them to the plugins directory for each node.
+        fun createDevMode(scanPackage: String): CordappLoader {
+            val resource = scanPackage.replace('.', '/')
+            val paths = javaClass.classLoader.getResources(resource)
+                    .asSequence()
+                    .map {
+                        println(it)
```
Comment (rnicoll) [inline node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt]: Old debugging code?

Comment (Clintonio) [reply node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt]: Ah yes, thanks,

File: node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt (hunk 1 above)
Comment (Clintonio) [reply node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt]: Done

File: node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+ * Handles CorDapp loading and classpath scanning
+ */
+class CordappLoader private constructor (val cordappClassPath: List<Path>) {
+    val appClassLoader: ClassLoader = javaClass.classLoader
+    val scanResult = scanCordapps()
+
+    companion object {
+        private val logger = loggerFor<CordappLoader>()
+
+        fun createDefault(baseDir: Path): CordappLoader {
```
Comment (rnicoll) [inline node/src/main/kotlin/net/corda/node/internal/classloading/CordappLoader.kt]: Not worth holding up this PR for, but could you document these two `create` functions please? Just a quick line like "Creates a CorDapp loader for normal usage. @param baseDir base directory of the node to search for plugins"...

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 5]
```diff
@@ -160,8 +153,18 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
         ServiceLoader.load(CordaPluginRegistry::class.java).toList()
     }
 
+    val cordappLoader: CordappLoader by lazy {
+        if (System.getProperty("net.corda.node.cordapp.scan.package") != null) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Extract the property value to a `val` as it's accessed twice

Comment (Clintonio) [reply node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: This has follow on work that is blocking this PR so I'll make sure it is in the next PR.

**Comment by shamsasari** [comment]
Oh, I hadn't finished my review. Shall I wait for your next PR?

File: node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt (hunk 1 above)
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/classloading/Utils.kt]: Can you move this to `InternalUtils.kt` instead

**Pull request merged**